### PR TITLE
Add sync test and refactor InParallel

### PR DESCRIPTION
**Pull request by tejal29**
Some work from #2091 

-  Added a test to make sure InSequence builds are run in sequence.
- Cleaned up parallel.go by adding functions.
- Used [sync.Map ](https://golang.org/pkg/sync/#Map)to collect parallel build results which are safe to use since as per the docs
``
The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex.
``
In our case, one go subroutine adds exactly one key and they are always disjoint. 
- Added Individual unit tests.
- Test cases cover code branches and not just errors.

### CODE REVIEW

File: pkg/skaffold/build/sequence_test.go [hunk 1]
```diff
@@ -85,3 +89,70 @@ func TestInSequence(t *testing.T) {
...
+
+			builder := concatTagger{}
+			got, err := InSequence(context.Background(), out, tags, artifacts, builder.doBuild)
+
+			testutil.CheckErrorAndDeepEqual(t, test.shouldErr, err, test.expected, got)
+		})
+	}
+}
+
+// summer builder sums all the numbers
```
Comment (balopat) [inline pkg/skaffold/build/sequence_test.go]:
```suggestion
// concatTagger sums all the numbers
```

File: pkg/skaffold/build/parallel.go [hunk 2]
```diff
@@ -33,84 +34,113 @@ const bufferedLinesPerArtifact = 10000
...
-		}
+	finalTag, err := getBuildResult(ctx, cw, tags, artifact, build)
+	if err != nil {
+		event.BuildFailed(artifact.ImageName, err)
+		results.Store(artifact.ImageName, err)
+	} else {
+		results.Store(artifact.ImageName, finalTag)
+	}
+
+	event.BuildComplete(artifact.ImageName)
```
[the remark this answers is not in the document]
Comment (tejal29) [reply pkg/skaffold/build/parallel.go]: It was the original behavior hence kept it here https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/build/parallel.go#L83

File: pkg/skaffold/build/parallel.go [hunk 3]
```diff
@@ -33,84 +34,113 @@ const bufferedLinesPerArtifact = 10000
...
+	for scanner.Scan() {
+		lines <- scanner.Bytes()
+	}
+	close(lines)
+}
+
+func setUpColorWriter(w io.WriteCloser, out io.Writer) io.WriteCloser {
+	var cw io.WriteCloser
+	if color.IsTerminal(out) {
+		cw = color.ColoredWriteCloser{WriteCloser: w}
```
[the remark this answers is not in the document]
Comment (tejal29) [reply pkg/skaffold/build/parallel.go]: done

File: pkg/skaffold/build/parallel.go [hunk 4]
```diff
@@ -33,84 +34,113 @@ const bufferedLinesPerArtifact = 10000
...
+		// Run build and write output/logs to piped writer and store build result in
+		// sync.Map
+		go runBuild(ctx, cw, tags, artifacts[i], results, buildArtifact)
+		// Read build output/logs and write to buffered channel
+		go readOutputAndWriteToChannel(r, outputs[i])
 	}
 
 	// Print logs and collect results in order.
-	var built []Artifact
+	printResults(out, outputs)
```
[the remark this answers is not in the document]
Comment (tejal29) [reply pkg/skaffold/build/parallel.go]: actually that is correct. Let me see if i can fix this.

[the remark this answers is not in the document]
Comment (tejal29) [reply pkg/skaffold/build/parallel.go]: @dgageot Please take another look. 
This is now fixed!

**Pull request merged**